### PR TITLE
chore: fix circular dependency between types and mutate-request

### LINKED ISSUE #1244: Remove circular dependencies
### Describe what should be investigated or refactored

Pepr has 18 circular dependencies. These can be the source of runtime errors, initialization order issues, and it complicates code structure. Furthermore, this makes things difficult to debug and can be the source of memory leaks or inconsistent behavior across environments.

We should restructure the code so that these circular dependencies are removed.

### Definition of Done

* There are less than 18 circular dependencies in the project.
* The tolerance threshold in CI/CD's `Circular Dependency Check` step is lowered.

Circular dependencies can be addressed over the course of several PRs if necessary, you do not need to resolve all of them at once.

### Additional context
Calculate the current amount of circular dependencies with: `npx madge -c --extensions ts src/`

Related issue to this broader issue are: 

```[tasklist]
### Tasks
- [ ] #1623
- [ ] #1624
- [ ] #1625
- [ ] #1626
- [ ] #1627
- [ ] #1628
- [ ] #1629
- [ ] #1630
- [ ] #1631
- [ ] #1632
- [ ] #1633
```

### PR DESCRIPTION
**Pull request by schaeferka**
## Description
We would like to remove the circular dependency between types.ts and mutate-request.ts
...

End to End Test:    
(See [Pepr Excellent Examples](https://github.com/defenseunicorns/pepr-excellent-examples))

## Related Issue

Fixes #

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging
- [ ] Unit, [Journey](https://github.com/defenseunicorns/pepr/tree/main/journey), [E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples), [docs](https://github.com/defenseunicorns/pepr/tree/main/docs), [adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or updated as needed
- [ ] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

### CODE REVIEW

File: .github/workflows/dependency-review.yml [hunk 1]
```diff
@@ -35,11 +35,11 @@ jobs:
...
           tail -n +4 tmp.log > circular-deps.log
-          if [ $(wc -l < circular-deps.log) -gt 18 ]; then
-            echo "circular-deps.log has more than 18 circular dependencies."
+          if [ $(wc -l < circular-deps.log) -gt 17 ]; then
+            echo "circular-deps.log has more than 17 circular dependencies."
             wc -l circular-deps.log
             exit 1
           else
-            echo "circular-deps.log has 18 or fewer circular dependencies."
+            echo "circular-deps.log has 17 or fewer circular dependencies."
```
Comment (samayer12) [inline .github/workflows/dependency-review.yml:43]: Lowering the tolerance threshold is something we need to do in addition to resolving the circular dep, as shown here.

File: src/lib/mutate-request.ts [hunk 2]
```diff
@@ -1,151 +1,10 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (samayer12) [inline src/lib/mutate-request.ts]: We shouldn't disable linter rules unless we have a really good reason. It looks like we can leave `PeprMutateRequest` and type `Mutate Action` in `mutate-request.ts` and extract the rest to `mutate-types.ts` as already shown. Let's see if that works before turning off that check.

File: src/lib/mutate-types.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+};
+
+// MutateAction type for handling mutation callbacks
+export type MutateAction<T extends GenericClass, K extends KubernetesObject = InstanceType<T>> = (
+  req: PeprMutateRequest<K>,
+  logger?: Logger,
+) => Promise<void> | void | Promise<PeprMutateRequest<K>> | PeprMutateRequest<K>;
+
+// PeprMutateRequest class for mutation request handling
+export class PeprMutateRequest<T extends KubernetesObject> {
```
Comment (samayer12) [inline src/lib/mutate-types.ts]: See other comment about leaving these in `mutate-request.ts`.

File: src/lib/types.ts [hunk 4]
```diff
@@ -2,20 +2,21 @@
...
+import { Operation } from "./mutate-types";
 import { WatchPhase } from "kubernetes-fluent-client/dist/fluent/types";
+import { Logger } from "pino";
 
-import { PeprMutateRequest } from "./mutate-request";
+import { PeprMutateRequest } from "./mutate-types";
 import { PeprValidateRequest } from "./validate-request";
 
-import { Logger } from "pino";
+//import { Logger } from "pino";
```
Comment (samayer12) [inline src/lib/types.ts]: Delete commented blocks

**Review by samayer12 — CHANGES_REQUESTED** [review]

File: .github/workflows/dependency-review.yml (hunk 1 above)
Comment (schaeferka) [reply .github/workflows/dependency-review.yml:43]: I'll make sure to lower the tolerance on future fixes

File: src/lib/types.ts (hunk 4 above)
Comment (schaeferka) [reply src/lib/types.ts]: Done

File: src/lib/mutate-request.ts (hunk 2 above)
Comment (schaeferka) [reply src/lib/mutate-request.ts]: Fixed with the changes you suggested

File: src/lib/mutate-types.ts (hunk 3 above)
Comment (schaeferka) [reply src/lib/mutate-types.ts]: Fixed with the changes you suggested

**Comment by schaeferka** [comment]
Closed and replaced with new PR to fix CI issue

**Pull request closed without merging**